portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -115,4 +159,39 @@ private String getParameter(HttpServletRequest request, String name, String defa
     }
     return value;
   }
+
+  /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // Use ImagesService to get a URL that points to the uploaded file.
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+
+    // To support running in Google Cloud Shell with AppEngine's dev server, we must use the relative
+    // path to the image, rather than the path returned by imagesService which contains a host.
+    try {
+      URL url = new URL(imagesService.getServingUrl(options));
+      return url.getPath();
+    } catch (MalformedURLException e) {
+      return imagesService.getServingUrl(options);
+    }
```

Comment:
Is this trying to verify whether the URL is valid (i.e. the image exists), and if so, why is the input URL returned in case of error?

Reply:
This was provided to us as a part of the starter code. From what i understand, url.getPath() will trim the scheme and the host (for example "https://google.com/?q=example" will become "/?q=example"). It's used to ensure that, on the dev server, ImageService can be used, since only relative paths will work on the dev server. https://github.com/thomasogara/step-repository/blob/b32c461af18067d17035fcb3a6c31a54c25676d0/walkthroughs/week-4-libraries/blobstore/examples/hello-world-fetch/src/main/java/com/google/sps/servlets/FormHandlerServlet.java#L64

Reply:
Ah I see, so it because of the differences between the dev server and the deployed server.

Reply:
Yep, that's it exactly.